clusterloader2/pkg/measurement/common/pod_startup_latency.go

```diff
@@ -0,0 +1,296 @@
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, p.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+	return nil
+}
+
+func (p *podStartupLatencyMeasurement) gather(c clientset.Interface) ([]measurement.Summary, error) {
+	glog.Infof("Gathering pod startup latency measurement...")
+	if p.informer == nil || p.stopCh == nil {
+		return []measurement.Summary{}, fmt.Errorf("metric %s has not been started", podStartupLatencyMetricName)
+	}
+
+	scheduleLag := make([]time.Duration, 0)
+	startupLag := make([]time.Duration, 0)
+	watchLag := make([]time.Duration, 0)
+	schedToWatchLag := make([]time.Duration, 0)
+	e2eLag := make([]time.Duration, 0)
+
+	close(p.stopCh)
+	if err := p.gatherScheduleTimes(c); err != nil {
+		return []measurement.Summary{}, err
+	}
+	for key, create := range p.createTimes {
+		sched, ok := p.scheduleTimes[key]
+		if !ok {
+			glog.Infof("Failed to find schedule time for %v", key)
+			continue
+		}
+		run, ok := p.runTimes[key]
+		if !ok {
+			glog.Infof("Failed to find run time for %v", key)
+			continue
+		}
+		watch, ok := p.watchTimes[key]
+		if !ok {
+			glog.Infof("Failed to find watch time for %v", key)
+			continue
+		}
+
+		scheduleLag = append(scheduleLag, sched.Time.Sub(create.Time))
+		startupLag = append(startupLag, run.Time.Sub(sched.Time))
+		watchLag = append(watchLag, watch.Time.Sub(run.Time))
+		schedToWatchLag = append(schedToWatchLag, watch.Time.Sub(sched.Time))
+		e2eLag = append(e2eLag, watch.Time.Sub(create.Time))
+	}
+
+	sort.Sort(durationSlice(scheduleLag))
+	sort.Sort(durationSlice(startupLag))
+	sort.Sort(durationSlice(watchLag))
+	sort.Sort(durationSlice(schedToWatchLag))
+	sort.Sort(durationSlice(e2eLag))
+
+	podStartupLatency := &podStartupLatency{
+		CreateToScheduleLatency: extractLatencyMetrics(scheduleLag),
+		ScheduleToRunLatency:    extractLatencyMetrics(startupLag),
+		RunToWatchLatency:       extractLatencyMetrics(watchLag),
+		ScheduleToWatchLatency:  extractLatencyMetrics(schedToWatchLag),
+		E2ELatency:              extractLatencyMetrics(e2eLag),
+	}
+	return []measurement.Summary{podStartupLatency}, nil
+}
+
+func (p *podStartupLatencyMeasurement) gatherScheduleTimes(c clientset.Interface) error {
+	selector := fields.Set{
+		"involvedObject.kind": "Pod",
+		"source":              corev1.DefaultSchedulerName,
+	}.AsSelector().String()
+	options := metav1.ListOptions{FieldSelector: selector}
+	schedEvents, err := c.CoreV1().Events(p.namespace).List(options)
+	if err != nil {
+		return err
+	}
+	for _, event := range schedEvents.Items {
+		key := createMetaNamespaceKey(event.InvolvedObject.Namespace, event.InvolvedObject.Name)
+		if _, ok := p.createTimes[key]; ok {
+			p.scheduleTimes[key] = event.FirstTimestamp
+		}
+	}
+	return nil
+}
+
+func (p *podStartupLatencyMeasurement) checkPodFunc() func(interface{}) {
+	var mutex sync.Mutex
```

Comment:
Why do you put this mutex here - it seems strange.

In general, you're trying to protect maps of p, so it should be mutex at the p level.

Reply:
Done.